libbeat/processors/add_process_metadata/add_process_metadata.go

```diff
@@ -0,0 +1,203 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_process_metadata
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/atomic"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+const (
+	processorName   = "add_process_metadata"
+	cacheExpiration = time.Second * 30
+)
+
+var (
+	// ErrNoMatch is returned when the event doesn't contain any of the fields
+	// specified in match_pids.
+	ErrNoMatch = errors.New("none of the fields in match_pids found in the event")
+
+	// ErrNoProcess is returned when metadata for a process can't be collected.
+	ErrNoProcess = errors.New("process not found")
+
+	procCache = newProcessCache(cacheExpiration, gosysinfoProvider{})
+
+	instanceID atomic.Uint32
+)
+
+type addProcessMetadata struct {
+	config   config
+	provider processMetadataProvider
+	log      *logp.Logger
+	mappings common.MapStr
+}
+
+type processMetadata struct {
+	name, title, exe string
+	args             []string
+	env              map[string]string
+	startTime        time.Time
+	pid, ppid        int
+	//
+	fields common.MapStr
+}
+
+type processMetadataProvider interface {
+	GetProcessMetadata(pid int) (*processMetadata, error)
+}
+
+func init() {
+	processors.RegisterPlugin(processorName, newProcessMetadataProcessor)
+}
+
+func newProcessMetadataProcessor(cfg *common.Config) (processors.Processor, error) {
+	return newProcessMetadataProcessorWithProvider(cfg, &procCache)
+}
+
+func newProcessMetadataProcessorWithProvider(cfg *common.Config, provider processMetadataProvider) (proc processors.Processor, err error) {
+	// Logging (each processor instance has a unique ID).
+	var (
+		id  = int(instanceID.Inc())
+		log = logp.NewLogger(processorName).With("instance_id", id)
+	)
+
+	config := defaultConfig()
+	if err = cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrapf(err, "fail to unpack the %v configuration", processorName)
+	}
+
+	p := addProcessMetadata{
+		config:   config,
+		provider: provider,
+		log:      log,
+	}
+	if p.mappings, err = config.getMappings(); err != nil {
+		return nil, errors.Wrapf(err, "error unpacking %v.target_fields", processorName)
+	}
+
+	return &p, nil
+}
+
+// Run enriches the given event with the host meta data
+func (p *addProcessMetadata) Run(event *beat.Event) (*beat.Event, error) {
+	for _, pidField := range p.config.MatchPIDs {
+		result, err := p.enrich(event.Fields, pidField)
+		if err != nil {
+			switch err {
+			case common.ErrKeyNotFound:
+				continue
+			case ErrNoProcess:
+				return event, err
+			default:
+				if !p.config.IgnoreErrors {
+					event = nil
```

Comment:
We don't yet have strong semantics around error handling in our processor pipeline. IIRC there are no other processors that discard events on error. So I would avoid returning a `nil` event.

If I configured the processor with `ignore_errors: false` then I would expect that either:
- the processor pipeline is cancelled and the event is sent without any additional processing (not possible today)
- the error is logged (default behavior when any processor returns an error)

If the user wants to discard the event then an alternative could be use a `drop_event` processor with a field exists conditional or we could add a `tag_on_failure` option here and the `drop_event` could be conditional on the tag.

Reply:
I'm thinking we need a "Processor Conventions" document somewhere to help ensure user's get a similar experience with each processor and that similar config options have the same name.

Reply:
++ on creating a convention and also what we do with `tags / flags`.